File: testSuite/benchmark-darkMatterOnlySubhalos.pl
#!/usr/bin/env perl
use strict;
use warnings;
use PDL;
use PDL::NiceSlice;
use PDL::Stats::Basic;
use JSON::PP;

# Run models to benchmark performance of a dark matter only subhalo evolution model.
# Andrew Benson (04-August-2022)

# Make output directory.
system("mkdir -p outputs/");

# Run the benchmark model multiple times.
my $runTimes = pdl [];
for(my $i=0;$i<11;++$i) {
    # Run the model.
    system("cd ..; /usr/bin/time --format=\"\%e\" --output=testSuite/outputs/benchmark_darkMatterOnlySubhalos.log ./Galacticus.exe testSuite/parameters/benchmark_darkMatterOnlySubHalos.xml");
    unless ( $? == 0 ) {
	print "FAIL: dark matter-only subhalos benchmark model failed to run\n";
	exit;
    }
    # Extract timing data.
    open(my $logFile,"outputs/benchmark_darkMatterOnlySubhalos.log");
    my $runTime = <$logFile>;
    close($logFile);
    chomp($runTime);
    $runTimes = $runTimes->append($runTime)
	unless ( $i == 0 );
}

# Find average and standard deviation of run times.
my $runTimeAverage           = $runTimes->average();
my $runTimeStandardDeviation = $runTimes->stdv   ();
print "Timings: ".$runTimes."\n";
print "Benchmark results: ".$runTimeAverage." ± ".$runTimeStandardDeviation." s\n";

# Generate JSON report.
my @output =
    (
     {
	 name  => "Dark Matter Only Subhalos - Wall Time",
	 unit  => "seconds"                              ,
	 value => $runTimeAverage          ->sclr()      ,
	 range => $runTimeStandardDeviation->sclr()
     }
    );

my $json = JSON::PP->new()->pretty()->encode(\@output);
open(my $reportFile,">","outputs/benchmark_darkMatterOnlySubhalos.json");
print $reportFile $json;
close($reportFile);

print "SUCCESS: dark matter-only subhalos benchmark model\n";

exit;

File: testSuite/benchmark-darkMatterOnlySubhalos.py
#!/usr/bin/env python3
import subprocess
import os
import json
import numpy as np

# Run models to benchmark performance of a dark matter only subhalo evolution model.
# Andrew Benson (04-August-2022; ported to Python)

# Make output directory.
os.makedirs("outputs", exist_ok=True)

# Run the benchmark model multiple times.
run_times = []
for i in range(11):
    # Run the model.
    result = subprocess.run(
        "cd ..; /usr/bin/time --format=\"%e\" --output=testSuite/outputs/benchmark_darkMatterOnlySubhalos.log ./Galacticus.exe testSuite/parameters/benchmark_darkMatterOnlySubHalos.xml",
        shell=True
    )
    if result.returncode != 0:
        print("FAIL: dark matter-only subhalos benchmark model failed to run")
        raise SystemExit(1)
    # Extract timing data (skip first run as warm-up).
    if i != 0:
        with open("outputs/benchmark_darkMatterOnlySubhalos.log") as f:
            run_times.append(float(f.read().strip()))

# Find average and standard deviation of run times.
run_times = np.array(run_times)
run_time_average       = run_times.mean()
run_time_standard_error = run_times.std() / np.sqrt(len(run_times))
print(f"Timings: {run_times}")
print(f"Benchmark results: {run_time_average} ± {run_time_standard_error} s")

# Generate JSON report.
output = [
    {
        "name":  "Dark Matter Only Subhalos - Wall Time",
        "unit":  "seconds",
        "value": float(run_time_average),
        "range": float(run_time_standard_error)
    }
]
with open("outputs/benchmark_darkMatterOnlySubhalos.json", "w") as f:
    json.dump(output, f, indent=4)

print("SUCCESS: dark matter-only subhalos benchmark model")
